include/fiction/algorithms/physical_design/graph_oriented_layout_design.hpp

```diff
@@ -0,0 +1,1574 @@
+    void foreach_gate_reverse(Fn&& fn) const
+    {
+        const uint32_t offset = 1u + this->num_pis() +
+                                (this->get_node(this->get_constant(true)) != this->get_node(this->get_constant(false)));
+        mockturtle::detail::foreach_element(topo_order.crbegin(), topo_order.crend() - offset, std::forward<Fn>(fn));
+    }
+
+    void update_topo()
+    {
+        this->incr_trav_id();
+        this->incr_trav_id();
+        topo_order.reserve(this->size());
+
+        /* constants and PIs */
+        const auto c0 = this->get_node(this->get_constant(false));
+        topo_order.push_back(c0);
+        this->set_visited(c0, this->trav_id());
+
+        if (const auto c1 = this->get_node(this->get_constant(true)); this->visited(c1) != this->trav_id())
+        {
+            topo_order.push_back(c1);
+            this->set_visited(c1, this->trav_id());
+        }
+
+        Ntk::foreach_ci([this](const auto& n) { create_topo_rec(n); });
+    }
+
+  private:
+    void create_topo_rec(const node& n)
+    {
+        /* is permanently marked? */
+        if (this->visited(n) == this->trav_id())
+        {
+            return;
+        }
+
+        bool not_visited = false;
+
+        this->foreach_fanin(n,
+                            [this, &not_visited](const signal& f)
+                            {
+                                if (this->visited(this->get_node(f)) != this->trav_id())
+                                {
+                                    not_visited = true;
+                                };
+                            });
+
+        if (not_visited)
+        {
+            return;
+        }
+
+        /* mark node n permanently */
+        this->set_visited(n, this->trav_id());
+
+        /* visit node */
+        topo_order.push_back(n);
+
+        /* mark children */
+        this->foreach_fanout(n, [this](const auto& fo) { create_topo_rec(fo); });
+    }
+
+    std::vector<node> topo_order;
+};
+/**
+ * Enum to specify which tiles should be checked if they are empty.
+ */
+enum class check_if_empty
+{
+    /**
+     * Do not check any tiles.
+     */
+    NONE,
+    /**
+     * Check if the source tile is empty.
+     */
+    SRC,
+    /**
+     * Check if the destination tile is empty.
+     */
+    DEST
+};
+/**
+ * Struct to hold information necessary for gate placement during layout generation for one vertex.
+ *
+ * @tparam ObstrLyt The type of the layout.
+ */
+template <typename ObstrLyt>
+struct placement_info
+{
+    /**
+     * The index of the current node being placed.
+     */
+    uint64_t current_node;
+    /**
+     * The index of the current primary output.
+     */
+    uint64_t current_po;
+    /**
+     * Mapping of nodes to their positions in the layout.
+     */
+    node_dict_type<ObstrLyt, tec_nt>& node2pos;
+    /**
+     * Mapping of primary input nodes to layout nodes.
+     */
+    mockturtle::node_map<mockturtle::node<ObstrLyt>, tec_nt>& pi2node;
```

Comment:
Is there a reason for using references here? This is error-prone since the respective variables can run out of scope. Dereferencing (accessing these members) will then cause a `SEGFAULT`. Better remove the `&` here. 

This is assuming that you're passing all instances of the `placement_info` object via reference anyway.